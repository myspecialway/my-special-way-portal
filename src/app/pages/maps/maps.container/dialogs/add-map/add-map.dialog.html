<div class="container add-map-modal">
  <h3 mat-dialog-title>הוספת מפה</h3>
  <mat-divider></mat-divider>

  <label for="map-name">שם מפה</label>
  <input type="text" id="map-name" class="form-control" [(ngModel)]="mapName" />

  <div ng2FileDrop class="upload-btn-wrapper"
    [ngClass]="{ 'nv-file-over': hasBaseDropZoneOver, 'file-uploaded': uploader.getNotUploadedItems().length }"
    [uploader]="uploader" (fileOver)="fileOverBase($event)" (click)="fileInput.click()">

    <button class="add-button">
      <i class="material-icons">insert_photo</i>
      הוסף תמונה מקובץ
    </button>

    <input type="file" #fileInput ng2FileSelect [uploader]="uploader"
      [disabled]="uploader.getNotUploadedItems().length || isSending"
      (change)="onChangeFile($event)"/>

    <div class="chosen-file">
      <strong>{{ uploader.queue[0]?.file?.name }}</strong>
      <span class="remove" (click)="removeFile($event)">x</span>
    </div>

    <app-spinner [isSpinnerVisible]="isSending"></app-spinner>
  </div>

  <mat-dialog-actions>
    <button class="mat-button" [disabled]="isSending" (click)="close()">ביטול</button>
    <button type="button" class="mat-raised-button mat-primary"
      [disabled]="!mapName || !uploader.getNotUploadedItems().length || isSending"
      (click)="sendFile()">
      הוסף
    </button>
  </mat-dialog-actions>
</div>
