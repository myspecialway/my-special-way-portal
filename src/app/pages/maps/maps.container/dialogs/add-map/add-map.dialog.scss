$purple-color: #8249d1;

* {
  direction: rtl;
  text-align: right;
}

.visible {
  visibility: visible;
}

::ng-deep .mat-dialog-container {
  border-radius: 4px;
  overflow: initial;
}

.add-map-modal {
  width: 310px;
}

.mat-dialog-title {
  font-family: Rubik;
  font-weight: 500;
  letter-spacing: 0.2px;
  margin: 0 0 10px;
}

.mat-divider {
  border-top: 3px solid $purple-color;
  margin-bottom: 28px;
}

label {
  color: #666666;
  font-size: 12px;
  margin-bottom: 0;
}

input {
  color: #222222;
  margin-bottom: 22px;
}

.upload-btn-wrapper {
  position: relative;
  overflow: hidden;
  font-size: 14px;
  margin-bottom: 18px;
  border: solid 3px transparent;

  &.nv-file-over:not(.file-uploaded) {
    border: dotted 3px $purple-color;
  }

  .add-button {
    color: $purple-color;
    background-color: white;
    border: 1px solid $purple-color;
    border-radius: 3px;
    margin-bottom: 15px;
    cursor: pointer;

    i {
      font-size: inherit;
      vertical-align: middle;
    }
  }

  .chosen-file {
    color: #191919;
    display: flex;
    justify-content: space-between;
    margin-bottom: 6px;
    padding: 0 10px;
    background-color: #f3f3f3;
    border-radius: 3px;
    letter-spacing: 1.3px;
    visibility: hidden;

    .remove {
      cursor: pointer;
    }
  }

  input[type='file'] {
    display: none;
  }

  &.file-uploaded {
    .add-button {
      opacity: 0.5;
      cursor: default;
    }

    .chosen-file {
      visibility: visible;
    }
  }
}

.mat-dialog-actions {
  display: flex;
  width: 100%;
  justify-content: flex-end;
  margin-bottom: -12px;

  button {
    text-align: center;
    font-family: Rubik;
    font-weight: 500;
    letter-spacing: 1.3px;
    border-radius: 3px;
  }
}
