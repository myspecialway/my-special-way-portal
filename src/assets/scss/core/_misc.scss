body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
