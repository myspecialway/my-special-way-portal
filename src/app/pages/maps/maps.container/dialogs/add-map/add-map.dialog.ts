import { Component, Inject, OnInit } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
import { FormGroup, FormControl, Validators, FormBuilder } from '@angular/forms';
import { FileUploader } from 'ng2-file-upload';

const URL = 'http://localhost:3000/api/upload';

@Component({
  selector: 'app-add-map.dialog',
  templateUrl: './add-map.dialog.html',
  styleUrls: ['./add-map.dialog.scss'],
})
export class AddMapDialogComponent implements OnInit {
  form: FormGroup;
  formControl = new FormControl('', [Validators.required]);
  uploader: FileUploader = new FileUploader({ url: URL, itemAlias: 'photo' });

  constructor(
    private formBuilder: FormBuilder,
    private dialogRef: MatDialogRef<AddMapDialogComponent>,
    @Inject(MAT_DIALOG_DATA) private data,
  ) {}

  ngOnInit(): void {
    this.form = this.formBuilder.group({
      mapName: '',
    });
  }

  getErrorMessage() {
    return this.formControl.hasError('required') ? 'Required field' : '';
  }

  close(): void {
    this.dialogRef.close();
  }

  sendAddMapRequest(dialogData) {
    this.dialogRef.close(dialogData);
    // TODO
  }
}
