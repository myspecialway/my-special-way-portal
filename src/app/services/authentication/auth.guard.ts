import gql from 'graphql-tag';
import { Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot, UrlSegment } from '@angular/router';
import { AuthenticationService } from './authentication.service';
import { UserType } from '../../models/user.model';
import { Apollo } from 'apollo-angular';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private router: Router,
    private authService: AuthenticationService,
    private apollo: Apollo,
  ) { }

  async canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
    await this.authService.checkRestoreAuthData();
    const response = await this.apollo.query<{ userProfile: UserProfileStateModel }>({
      query: gql`
      query {
        userProfile @client{
          role
          token
        }
      }
    `}).toPromise();

    if (
      response.data.userProfile &&
      response.data.userProfile.token &&
      !this.authService.isTokenExpired(response.data.userProfile.token)
      && this.isAuthorized(route.url, response.data.userProfile.role)
    ) {
      return true;
    }

    this.router.navigate(['/login'], { queryParams: { returnUrl: state.url } });
    return false;
  }

  isAuthorized(route: UrlSegment[], userRole: UserType) {
    // console.log(`route ${route} | User type ${userRole}`);
    // if (route[0].path.toString() === 'user' && userRole.toString() === 'TEACHER') {
    //   return false;
    // }
    return true; // TODO: authorization logic
  }
}
