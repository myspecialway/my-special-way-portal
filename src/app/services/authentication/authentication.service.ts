import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { LoginResponse } from '../../models/login-response.model';
import { environment } from '../../../environments/environment';
import { JwtHelperService } from '@auth0/angular-jwt';
import { JWTTokenPayloadResponse } from '../../models/jwt-token-resonse.model';
import { Apollo } from 'apollo-angular';
import { UPDATE_USER_PROFILE } from '../../apollo/state/mutations/update-user-profile.mutation';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';
import { UserUniqueValidationResponse } from '../../models/user-unique-validation-response.model';
import { of } from 'rxjs';
import { catchError, first, map } from 'rxjs/operators';

@Injectable()
export class AuthenticationService {
  private jwtHelper = new JwtHelperService();
  private isStateRestored = false;

  constructor(private http: HttpClient, private apollo: Apollo) {}

  private getTokenFromLocalStore() {
    return localStorage.getItem('token') ? localStorage.getItem('token') : sessionStorage.getItem('token');
  }

  async checkRestoreAuthData() {
    if (this.isStateRestored) {
      return;
    }

    this.isStateRestored = true;
    const token = this.getTokenFromLocalStore();

    if (!token) {
      return;
    }

    const userProfile = this.getProfileFromToken(token);
    await this.pushUserProfileToState(userProfile);
  }

  async login(username: string, password: string, isRememberMeActive: boolean): Promise<boolean> {
    try {
      const tokenResponse = await this.http
        .post<LoginResponse>(environment.hotConfig.MSW_HOT_LOGIN_ENDPOINT, { username, password })
        .toPromise();

      if (!tokenResponse) {
        return false;
      }

      if (isRememberMeActive) {
        localStorage.setItem('token', tokenResponse.accessToken);
      } else {
        sessionStorage.setItem('token', tokenResponse.accessToken);
      }

      const userProfile = this.getProfileFromToken(tokenResponse.accessToken);
      await this.pushUserProfileToState(userProfile);

      return true;
    } catch (error) {
      const typedError = error as HttpErrorResponse;

      if (typedError.status !== 401) {
        throw error;
      }

      return false;
    }
  }

  private getProfileFromToken(token: string): UserProfileStateModel {
    const jwrParsedToken = this.parseToken(token);
    const userProfile = new UserProfileStateModel(jwrParsedToken);
    userProfile.token = token;

    return userProfile;
  }

  private parseToken(token: string): JWTTokenPayloadResponse {
    const decodedToken = this.jwtHelper.decodeToken(token) as JWTTokenPayloadResponse;
    return decodedToken;
  }

  private async pushUserProfileToState(userProfile: UserProfileStateModel) {
    return this.apollo
      .mutate({
        mutation: UPDATE_USER_PROFILE,
        variables: {
          userProfile,
        },
      })
      .toPromise();
  }

  isTokenExpired(token: string): boolean {
    return this.jwtHelper.isTokenExpired(token);
  }

  logout() {
    localStorage.removeItem('token');
    sessionStorage.removeItem('token');
  }

  checkUsernameUnique(username: string, id: string) {
    if (!username) {
      return of(true);
    }
    const userID = id === '' ? undefined : id;
    try {
      return this.http
        .post<UserUniqueValidationResponse>(environment.hotConfig.MSW_HOT_USER_UNIQUE_VALIDATION_ENDPOINT, {
          username,
          id: userID,
        })
        .pipe(
          first(),
          map((res) => res.isUnique),
          catchError(() => of(true)),
        );
    } catch (error) {
      return of(true);
    }
  }
}
