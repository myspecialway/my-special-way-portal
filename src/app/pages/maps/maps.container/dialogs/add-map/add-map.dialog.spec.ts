import { MatDialogRef, MatDialogModule } from '@angular/material';
import { TestBed, ComponentFixture } from '@angular/core/testing';
import { AddMapDialogComponent } from './add-map.dialog';
import { NO_ERRORS_SCHEMA } from '@angular/compiler/src/core';
import { AuthenticationService } from '../../../../../services/authentication/authentication.service';
import { HttpClient, HttpHandler } from '@angular/common/http';
import { Apollo } from 'apollo-angular';
import { FileUploadModule } from 'ng2-file-upload';

describe('AddMapDialogComponent', () => {
  let authServiceMock: Partial<AuthenticationService>;
  let mockDialogRef: any;
  beforeEach(() => {
    authServiceMock = {
      getTokenFromLocalStore: jest.fn(),
    };
    mockDialogRef = {
      close: jest.fn(),
    };
  });
  beforeEach(async () => {
    TestBed.configureTestingModule({
      imports: [MatDialogModule, FileUploadModule],
      declarations: [AddMapDialogComponent],
      providers: [
        Apollo,
        HttpClient,
        HttpHandler,
        AuthenticationService,
        {
          provide: MatDialogRef,
          useValue: mockDialogRef,
        },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  });

  let fixture: ComponentFixture<AddMapDialogComponent>;
  let component: AddMapDialogComponent;

  beforeEach(() => {
    fixture = TestBed.createComponent(AddMapDialogComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
