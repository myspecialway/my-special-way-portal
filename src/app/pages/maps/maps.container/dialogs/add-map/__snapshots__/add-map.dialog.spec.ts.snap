// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMapDialogComponent should match snapshot 1`] = `
<app-add-map
  authenticationService={[Function AuthenticationService]}
  dialogRef={[Function Object]}
  isSending="false"
  mapName=""
  uploader={[Function FileUploader]}
>
  <div
    class="container add-map-modal"
  >
    <h3
      class="mat-dialog-title"
      id="mat-dialog-title-0"
      mat-dialog-title=""
    >
      הוספת מפה
    </h3>
    <mat-divider />
    <label
      for="map-name"
    >
      שם מפה
    </label>
    <input
      class="form-control"
      id="map-name"
      type="text"
    />
    <div
      class="upload-btn-wrapper"
      ng-reflect-klass="upload-btn-wrapper"
      ng-reflect-ng-class="[object Object]"
      ng-reflect-uploader="[object Object]"
      ng2filedrop=""
    >
      <button
        class="add-button"
      >
        <i
          class="material-icons"
        >
          insert_photo
        </i>
         הוסף תמונה מקובץ 
      </button>
      <input
        ng-reflect-uploader="[object Object]"
        ng2fileselect=""
        type="file"
      />
      <div
        class="chosen-file"
      >
        <strong>
          
        </strong>
        <span
          class="remove"
        >
          x
        </span>
      </div>
      <app-spinner />
    </div>
    <mat-dialog-actions
      class="mat-dialog-actions"
    >
      <button
        class="mat-button"
      >
        ביטול
      </button>
      <button
        class="mat-raised-button mat-primary"
        disabled=""
        type="button"
      >
         הוסף 
      </button>
    </mat-dialog-actions>
  </div>
</app-add-map>
`;
